### ISSUE TITLE
Repl evaluation delayed by one step

### ISSUE THREAD

**Issue by olof3** [timholy/Revise.jl#647]
This has come up before, e.g., #195, but it seemed difficult to reproduce. Not sure if it got tracked down, if this is something else, or if it shouldn't be expected to work.

On my machine I can reproduce the problem with the code below. After `using ExponentialUtilities`, I ran `]add ExponentialUtilities@v1.9.0` in _another_ REPL (doing it in the same REPL it  works alright).

```               _
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.8.0-DEV.480 (2021-09-08)
 _/ |\__'_|_|_|\__'_|  |  Commit 62f0ccda46* (16 days old master)
|__/                   |

(@v1.8) pkg> status
Status `~/.julia/environments/v1.8/Project.toml`
 [6e4b80f9] BenchmarkTools v1.1.4
 [d4d017d3] ExponentialUtilities v1.10.0
 [295af30f] Revise v3.1.19
 [8bb1440f] DelimitedFiles

julia> using ExponentialUtilities

julia> 2+2
ERROR: /home/luke/.julia/packages/ExponentialUtilities/DdWzW/src/ExponentialUtilities.jl is not stored in the source-text cache
Stacktrace:
  [1] error(::String, ::String)
    @ Base ./error.jl:42
  [2] _read_dependency_src(io::IOStream, filename::String)
    @ Base ./loading.jl:1637
  [3] read_dependency_src(io::IOStream, filename::String)
    @ Base ./loading.jl:1623
  [4] read_dependency_src(cachefile::String, filename::String)
    @ Base ./loading.jl:1644
  [5] read_from_cache(pkgdata::Revise.PkgData, file::String)
    @ Revise ~/.julia/packages/Revise/OgnOk/src/pkgs.jl:109
  [6] maybe_parse_from_cache!(pkgdata::Revise.PkgData, file::String)
    @ Revise ~/.julia/packages/Revise/OgnOk/src/pkgs.jl:119
  [7] (::Revise.var"#44#45"{String})()
    @ Revise ~/.julia/packages/Revise/OgnOk/src/pkgs.jl:526
  [8] with_logstate(f::Function, logstate::Any)
    @ Base.CoreLogging ./logging.jl:511
  [9] with_logger
    @ ./logging.jl:623 [inlined]
 [10] watch_manifest(mfile::String)
    @ Revise ~/.julia/packages/Revise/OgnOk/src/pkgs.jl:505
 [11] (::Revise.TaskThunk)()
    @ Revise ~/.julia/packages/Revise/OgnOk/src/types.jl:272

julia> 1+1
4

julia> 2+2
2

(@v1.8) pkg> status
Status `~/.julia/environments/v1.8/Project.toml`
 [6e4b80f9] BenchmarkTools v1.1.4
 [d4d017d3] ExponentialUtilities v1.9.0
 [295af30f] Revise v3.1.19
 [8bb1440f] DelimitedFiles
```

**Comment by timholy** [comment]
Thanks for filing this. I've seen reports of this lag, but I've never been able to get a reproducer.

Sadly, I can't replicate this; it never throws the error for me. Can I ask you to try #649? (`pkg> add Revise#teh/fix_647` should do the trick).

**Comment by olof3** [comment]
Thanks, I tried the fix, but the problems persists on my machine. This time I also tried with a different package and without `startup.jl`,just in case.

```
$ julia/julia --startup-file=no
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.8.0-DEV.480 (2021-09-08)
 _/ |\__'_|_|_|\__'_|  |  Commit 62f0ccda46* (18 days old master)
|__/                   |

julia> using Revise

(@v1.8) pkg> status
Status `~/.julia/environments/v1.8/Project.toml`
 [d360d2e6] ChainRulesCore v1.6.0
 [295af30f] Revise v3.1.19 `https://github.com/timholy/Revise.jl.git#teh/fix_647`

julia> using ChainRulesCore

julia> 2+2
ERROR: /home/luke/.julia/packages/ChainRulesCore/ChM7X/src/compat.jl is not stored in the source-text cache
Stacktrace:
  [1] error(::String, ::String)
    @ Base ./error.jl:42
  [2] _read_dependency_src(io::IOStream, filename::String)
    @ Base ./loading.jl:1637
  [3] read_dependency_src(io::IOStream, filename::String)
    @ Base ./loading.jl:1623
  [4] read_dependency_src(cachefile::String, filename::String)
    @ Base ./loading.jl:1644
  [5] read_from_cache(pkgdata::Revise.PkgData, file::String)
    @ Revise ~/.julia/packages/Revise/9MkNl/src/pkgs.jl:109
  [6] maybe_parse_from_cache!(pkgdata::Revise.PkgData, file::String)
    @ Revise ~/.julia/packages/Revise/9MkNl/src/pkgs.jl:119
  [7] (::Revise.var"#44#45"{String})()
    @ Revise ~/.julia/packages/Revise/9MkNl/src/pkgs.jl:526
  [8] with_logstate(f::Function, logstate::Any)
    @ Base.CoreLogging ./logging.jl:511
  [9] with_logger
    @ ./logging.jl:623 [inlined]
 [10] watch_manifest(mfile::String)
    @ Revise ~/.julia/packages/Revise/9MkNl/src/pkgs.jl:505
 [11] (::Revise.TaskThunk)()
    @ Revise ~/.julia/packages/Revise/9MkNl/src/types.jl:272

julia> 1+1
4

(@v1.8) pkg> status
Status `~/.julia/environments/v1.8/Project.toml`
 [d360d2e6] ChainRulesCore v1.5.0
 [295af30f] Revise v3.1.19 `https://github.com/timholy/Revise.jl.git#teh/fix_647`
```

**Comment by timholy** [comment]
Thanks for checking! I was initially surprised but now it makes sense. Can you update that branch in #649 (including the "stop adding to response_channel" commit) and try again?

**Comment by olof3** [comment]
Now the REPL lag seems to be gone :+1: 

Not sure if the revising is taking place as expected, and how much that should be expected. Changing the version of `ChainRulesCore` does not give an error but for `ExponentialUtilities` v1.9 -> v1.10 it does, I guess it could be due that changes to the file names? Also not all methods in `v1.10` show up, e.g., `exponential!`

```
luke@luke-ThinkPad-T490s:~$ julia/julia --startup-file=no
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.8.0-DEV.480 (2021-09-08)
 _/ |\__'_|_|_|\__'_|  |  Commit 62f0ccda46* (18 days old master)
|__/                   |

julia> using ChainRulesCore

julia> using Revise

(@v1.8) pkg> status
Status `~/.julia/environments/v1.8/Project.toml`
 [d360d2e6] ChainRulesCore v1.6.0
 [d4d017d3] ExponentialUtilities v1.9.0
 [295af30f] Revise v3.1.19 `https://github.com/timholy/Revise.jl.git#teh/fix_647`

julia> 2+2
4

(@v1.8) pkg> status
Status `~/.julia/environments/v1.8/Project.toml`
 [d360d2e6] ChainRulesCore v1.5.0
 [d4d017d3] ExponentialUtilities v1.9.0
 [295af30f] Revise v3.1.19 `https://github.com/timholy/Revise.jl.git#teh/fix_647`

julia> using ExponentialUtilities

julia> ┌ Error: Error watching manifest
│   exception =
│    /home/luke/.julia/packages/ExponentialUtilities/Uk7jY/src/exp.jl is not stored in the source-text cache
│    Stacktrace:
│     [1] error(::String, ::String)
│       @ Base ./error.jl:42
│     [2] _read_dependency_src(io::IOStream, filename::String)
│       @ Base ./loading.jl:1637
│     [3] read_dependency_src(io::IOStream, filename::String)
│       @ Base ./loading.jl:1623
│     [4] read_dependency_src(cachefile::String, filename::String)
│       @ Base ./loading.jl:1644
│     [5] read_from_cache(pkgdata::Revise.PkgData, file::String)
│       @ Revise ~/.julia/packages/Revise/vo4CO/src/pkgs.jl:109
└ @ Revise ~/.julia/packages/Revise/vo4CO/src/pkgs.jl:542
julia> ExponentialUtilities.
@diagview                     _phiv_timestep_estimate_flops  exp_pade_p!                    kiops
ExpvCache                     _rdiv!                         exp_pade_q                     kiops_update_solution!
HermitianSubspaceCache        applyA!                        expv                           lanczos!
KrylovSubspace                arnoldi                        expv!                          lanczos_step!
PhivCache                     arnoldi!                       expv_timestep                  ldiv_for_generated!
Stegr                         arnoldi_step!                  expv_timestep!                 naivemul!
StegrCache                    checkdims                      firststep!                     phi
SubspaceCache                 coeff                          getH                           phi!
__init__                      eval                           getV                           phiv
_baseexp!                     expT!                          get_cache                      phiv!
_const                        exp_gen!                       get_caches                     phiv_dense
_exp!                         exp_generic                    get_subspace_cache             phiv_dense!
_expv_ee                      exp_generic!                   getmem                         phiv_timestep
_expv_hb                      exp_generic_core!              include                        phiv_timestep!
_phiv_timestep_adapt          exp_generic_mutable            inplace_add!                   realview
_phiv_timestep_caches         exp_pade_p                     intlog2
(@v1.8) pkg> status
Status `~/.julia/environments/v1.8/Project.toml`
 [d360d2e6] ChainRulesCore v1.5.0
 [d4d017d3] ExponentialUtilities v1.10.0
 [295af30f] Revise v3.1.19 `https://github.com/timholy/Revise.jl.git#teh/fix_647`
```

**Comment by timholy** [comment]
Progress! Does it get even better if you wrap https://github.com/timholy/Revise.jl/blob/3136226d6491098627b28f9feb002a95678f0664/src/pkgs.jl#L526
in
```julia
try
    maybe_extract_sigs!(maybe_parse_from_cache!(pkgdata, file))
catch err
    push!(revision_errors, (pkgdata, file))
    queue_errors[(pkgdata, file)] = (err, catch_backtrace())
    continue
end
```
Since this is hard to test I'm kinda guessing there, but that shouldn't be too far off. If this doesn't work try commenting out the `push!(revision_errors...)` and or `queue_errors` lines and see if that gets better.

**Comment by olof3** [comment]
Just doing the change gives
```
julia> ┌ Error: Error watching manifest
│   exception =
│    UndefVarError: revision_errors not defined
│    Stacktrace:
│     [1] (::Revise.var"#44#45"{String})()
│       @ Revise ~/.julia/packages/Revise/vo4CO/src/pkgs.jl:529
```
I guess it is a typo, but I'm not sure what to replace `revision_errors` with.

If I comment out `push!(revision_errors, (pkgdata, file))` I no longer see an error, but functions only present in the newer package version does not get revised in.

**Comment by timholy** [comment]
Sorry, that was supposed to be `revision_queue`. It mimics https://github.com/timholy/Revise.jl/blob/3136226d6491098627b28f9feb002a95678f0664/src/packagedef.jl#L797-L798

**Comment by timholy** [comment]
Is #649 better now? Note there's a new test which mimics what you're seeing. Unfortunately I've never seen the `is not stored in the source-text cache` so this likely still suffers from problems. But for me the new test passes.

**Comment by olof3** [comment]
With the updated #649 I get an other error. Not sure if I used `active_project_watchers()` the right way. It seemed a bit involved to use the dev version of Pkg, so I skipped that, but I'd be happy to do so if necessary.

```               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.8.0-DEV.480 (2021-09-08)
 _/ |\__'_|_|_|\__'_|  |  Commit 62f0ccda46* (19 days old master)
|__/                   |

julia> using Revise
[ Info: isdefined: false

julia> Revise.active_project_watcher()

julia> using ExponentialUtilities

julia> Revise.LogRecord(Error, error parsing cache for ~/.julia/packages/ExponentialUtilities/DdWzW : src/ExponentialUtilities.jl, pkgs, Revise_78247a5e, "~/.julia/packages/Revise/7mecm/src/pkgs.jl", 531, ()~/.julia/packages/ExponentialUtilities/Uk7jY/src/ExponentialUtilities.jl is not stored in the source-text cache
Stacktrace:
 [1] error(::String, ::String)
   @ Base ./error.jl:42
 [2] _read_dependency_src(io::IOStream, filename::String)
   @ Base ./loading.jl:1637
 [3] read_dependency_src(io::IOStream, filename::String)
   @ Base ./loading.jl:1623
 [4] read_dependency_src(cachefile::String, filename::String)
   @ Base ./loading.jl:1644
 [5] read_from_cache(pkgdata::Revise.PkgData, file::String)
   @ Revise ~/.julia/packages/Revise/7mecm/src/pkgs.jl:109)
```
the last error repeats some 10 times.

**Comment by timholy** [comment]
#649 is as close to a fix as it's currently possible to achieve. The REPL lag is gone, and under some circumstances Revise works as you'd hope.

However, there is a serious issue that needs to be fixed in Julia itself (that's what the "circumstances" are about). It turns out that while we save multiple versions of the package source, the `*.ji` cache file name is *not* dependent on the package version. Hence if in another process you update from 1.9->1.10 (or vice versa), it can overwrite the `*.ji` file before Revise has a chance to parse the source. (Revise extracts the old source from the cache file to avoid having to re-parse all your code defensively.)

There are two "fixes" currently:
- Revise could call `maybe_parse_from_cache!` immediately on each file after loading, so as to defensively parse everything. I can promise you, no one will like this solution, but it would be bulletproof.
- you can set `ENV["JULIA_PKG_PRECOMPILE_AUTO"]=0` to disable automatic precompilation when you change versions. This does not really fix the problem, but it makes it less likely to occur.

However, I think the real fix is to ensure that the old cache file doesn't get overwritten.

**Comment by timholy** [comment]
This also fixed #645 and #617, so by sticking with me here I think we've made a pretty big improvement. Thanks very much, @olof3!
